### ISSUE TITLE
Strong params deprecated with double splat operators

### ISSUE THREAD

**Issue by basawyer** [rails/rails#26201]
The move away from HWIA inheritance in ActionController::Parameters is the right move; strong parameters are not hashes. However, they should still work with the double splat operators and named parameters. I was actually surprised to see AR uses `keys` to save attributes in strong parameters.  Either way, double splatting strong parameters is currently deprecated. Example:

```
def meth(a: nil, b: nil)
  puts a, b
end

def class_create_method(**args)
  meth(**args)
end

# Controller
permitted_params = ActionController::Parameters.new({ a: 1, b: 2 }).permit(:a,:b)

class_create_method(permitted_params)

# DEPRECATION WARNING: Method to_hash….

```

I'm putting together a pull request now.

**Rails version**: 5.0.0.1

**Ruby version**: 2.3.1

**Comment by rafaelfranca** [comment]
While the deprecation can be fixed this never worked and will not, neither in Rails 4.2.

```
[1] pry(main)> def meth(a: nil, b: nil)
[1] pry(main)*   puts a, b
[1] pry(main)* end
=> :meth
[2] pry(main)> def class_create_method(**args)
[2] pry(main)*   meth(**args)
[2] pry(main)* end
=> :class_create_method
[3] pry(main)> permitted_params = ActionController::Parameters.new({ a: 1, b: 2 }).permit(:a,:b)
=> {"a"=>1, "b"=>2}
[4] pry(main)> class_create_method(permitted_params)
ArgumentError: wrong number of arguments (1 for 0)
from (pry):4:in `class_create_method'
[5] pry(main)> **permited_params
SyntaxError: unexpected **arg, expecting end-of-input
**permited_params
  ^
[5] pry(main)> class_create_method(a: 1)
```

**Comment by amilligan** [comment]
@rafaelfranca I think the fundamental problem here lies with HWIA, more than ActionController::Parameters.  Specifically, `**` does work with any object that responds to `#to_hash` -- which ActionController::Parameters, of course, does in Rails 4.x.  

Here's an addition to your examples:

```ruby
[6] pry(main)> { **permited_params }
TypeError: wrong argument type String (expected Symbol)
```

The problem here is not that AC::Params doesn't respond to `**`, but that it does (in 4.x) respond to `#to_hash`.  The `**` invokes the implicit conversion, but the result of that conversion is a hash with string keys.  I don't know why HWIA converts keys to strings -- I expect the decision to do so is lost to the mists of time -- but ruby wants hashes to have symbol keys.  So, this will work (in 4.x):

```ruby
[7] pry(main)> { **permited_params.symbolize_keys }
```

There's no immediate reason for it not to continue to work, assuming `#symbolize_keys` (and probably whatever other hash methods are currently deprecated in 5.0) delegate to `#to_h`.

We could start a debate on whether a HWIA should subclass Hash -- i.e. can we say a HashWithIndifferentAccess IS A Hash?  To wit:

```ruby
> { **{ a: 1, b: 2 } }
=> {:a=>1, :b=>2}

> { **{ a: 1, b: 2 }.with_indifferent_access }
TypeError: wrong argument type String (expected Symbol)
```

but perhaps that is an issue for another time.

**Comment by rafaelfranca** [comment]
Parameters are fundamentally hashes with string keys, not matter if we use HWIA as storage, they are instantiated with string keys since this is what we get from the request.

I'm totally fine to keep `** permited_params.deep_symbolize_keys` working, what will never happen is us doing the deep symbolizing by default in `to_hash`.

> We could start a debate on whether a HWIA should subclass Hash -- i.e. can we say a HashWithIndifferentAccess IS A Hash?

Yes, we can. It is a hash with keys stored as strings.

```
>> { **{ "a" => 1 } }
TypeError: wrong argument type String (expected Symbol)
```

> I don't know why HWIA converts keys to strings

We need to normalize the storage of the hash to one type if we need indifferent access but don't want the performance overhead of the conversion on reading the value. We could both convert keys to symbol or convert keys to string. Doing `to_sym` conversion in Ruby < 2.2 means that symbol can't be GCed, so imagine doing that with parameters of a request, you would need infinity memory. Now we can change the storage to be symbols but I'm sure that will break some applications and this breaking change is not worth.

**Comment by amilligan** [comment]
> > We could start a debate on whether a HWIA should subclass Hash -- i.e. can we say a HashWithIndifferentAccess IS A Hash?
> 
> Yes, we can. It is a hash with keys stored as strings.
You and I both know that's not how Liskov Substitutability works.

I'm not suggesting we symbolize keys by default; I had a conversation with @basawyer a few days ago, and I may have led him astray.  We can drop that from the conversation.

I believe the underlying goal of this patch is to restore the behavior of a HWIA to ActionController::Parameters; this should be possible with no security implications now that the subclassing relationship has been eliminated.  We can simply delegate Hash methods to the result of `#to_h`.  I'll talk with @basawyer about submitting a new patch within a day or so.

**Comment by basawyer** [comment]
@rafaelfranca @amilligan I've updated my pull request #26203.

**Comment by kamipo** [comment]
Fixed by #28734.
